Avatar privacy and new scope

PVince81 opened the pull request:
### Todos

- [x] Private avatars are now only shown to their owners.
- [x] Avatar that have not been set to public will only be shown for logged in users.
- [x] New option for "v2-private" but
   - [x] Add option in menu
   - [x] Make sure the option is working in various places
   - [x] Only make the extra option available for "Avatar", "Phone number", "Twitter", "Website", "Address"
   - [x] Scope menu must always be visible even when federation / public servers option is off
   - [x] Remove federated options when lookup server is disabled
   - [x] Unit tests
- [x] extend OCS provisioning endpoint "https://vvortex.local/ocs/v2.php/cloud/users/$userId" 
   - [x] allow getting scope values
   - [x] allow setting the scope values
   - [x] allow setting other values like phone, etc (if not already)
   - [x] users can get/set fields and scope values for themselves, but not for other users
   - [x] admin **cannot** get/set fields for other users
   - [x] Unit tests
- [x] add boolean capability to signal the ability to set scopes

### Issues

- [x] Issue? Switching avatar privacy doesn't reset cache for browsers (Chromium doesn't reload avatars on shift+refresh) => inform that it's only about 1 day caching
- [x] It seems the wording of the current options is already confusing, as they are about federated sharing => solved
- [x] Improve performance of placeholder avatar (currently using the uncached guest avatar)
- [x] Phone, address, twitter, website fields need to be always visible and editable (never disabled)
- [ ] What icons to use for "Private" and "Local" ?
- [x] Test/check avatar endpoint with mobile clients as they might use the public unauthenticated endpoint and raise tickets accordingly => I was told that they use authentication headers
- [x] Fix AvatarManagerTest

PVince81 committed:
Add new v2-private account scope

Added new v2-private account manager scope that restricts the scope
further by excluding public link access.

Avatars with v2-private account scope are now showing the guest avatar
instead of the real avatar.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Adjust scopes menu based on conditions

Now not all fields have the "v2-private" option in place.
Fix dropdown issue when a scope was stored that is not listed after
disabling the lookup server.
Whenever the lookup server upload is disabled, the scope menu is now
displayed where it makes sense to allow switching between the two private
scopes.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Map old account scope properties to new names

Use new scope values in settings page.
Adjust all consumers to use the new constants.
Map old scope values to new ones in account property getter.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Make extra user profile fields always editable

The fields for phone number, address, website and twitter are now
editable regardless whether federated sharing and the lookup server
are enabled or not.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
- [ ] BUG: migrating from old values seems to not work

PVince81 commented:
Updated top post to also extend the OCS API endpoint to allow clients to get/set scope values

PVince81 commented:
hah... it appears that the provisioning API is also cheating and using AccountManager->getUser(), so the mapping will not work there. I'll migrate it to use `IAccountManager->getAccount()` as part of the OCS API extension.

PVince81 committed:
OCS allow reading and writing account property scopes

Extends the provisioning API to allow a user to get and set their own
account property scopes.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
OCS API has been adjusted.

Next up: unit tests

nickvergessen commented on apps/provisioning_api/lib/Controller/AUserData.php:
```diff
@@ -138,11 +140,26 @@ protected function getUserData(string $userId): array {
+			IAccountManager::PROPERTY_PHONE,
+			IAccountManager::PROPERTY_ADDRESS,
+			IAccountManager::PROPERTY_WEBSITE,
+			IAccountManager::PROPERTY_TWITTER,
```
avatar can toggle between scopes private and local in the future

PVince81 replied:
indeed, I was so focused on those fields that I forgot avatar.

I'll add it everywhere where applicable

PVince81 replied:
added

PVince81 committed:
Expose avatarScope through provisioning API

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Enhance UsersControllerTest of provisioning API with scopes

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Add property scope tests for AccountManager

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
unit tests done

next up would be fixing guest avatar performance, but could also be done separately as the current approach already works

however, still need to settle at least on wording and icons

PVince81 commented:
Added task "add capability". I think a simple boolean should do to signal the ability to read and set scopes.

PVince81 commented:
- [ ] switch icons to material design ? (so that clients could use the same icon names from the MD library)

PVince81 committed:
Add capability for editable scopes in provisioning API

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
maximum storage wastage per user for placeholder avatars: 38 Megabytes!

PVince81 committed:
Added PlaceholderAvatar with own cached images

When avatar scope is private, the PlaceholderAvatar is used to deliver a
placeholder avatar based on the user's initials.

This was implemented as a separate class for now to avoid messing with
the existing UserAvatar implementation and its generated vs
non-generated logic.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
Added capability:
```
        "provisioning_api": {
          "version": "1.11.0",
          "hasAccountPropertyScopes": true
        },
```

Added `PlaceholderAvatar` which is like `GuestAvatar` but caches the images.
This was done separately to avoid messing with existing UserAvatar logic and their own existing generated images, which would require more logic and migration to make it work right.

PVince81 committed:
Update psalm-baseline for Avatar API quirks

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

nickvergessen commented on apps/provisioning_api/lib/Capabilities.php:
```diff
@@ -0,0 +1,51 @@
+		return [
+			'provisioning_api' => [
+				'version' => $this->appManager->getAppVersion('provisioning_api'),
+				'hasAccountPropertyScopes' => true,
```
```suggestion
				'hasAccountPropertyScopesV2' => true,
```
or
```suggestion
				'AccountPropertyScopesVersion' => 2,
```
Or something?

nickvergessen commented on lib/private/Avatar/AvatarManager.php:
```diff
@@ -104,12 +117,30 @@ public function getAvatar(string $userId) : IAvatar {
 		}
 
+		// requesting in public page
+		if ($requestingUser === null) {
```
```suggestion
		if ($requestingUser === null || !$this->knownUsersService->isKnownToUser($requestingUser->getUID(), $userId);) {
```

PVince81 replied:
I'm already on it with a slightly different code path, we need to make sure $requestingUser is not null and also $requestingUser !== $userId.

does the service consider self as known ?

nickvergessen replied:
Well "$requestingUser is not null" is handled by the part before the `||` so you won't reach it.
But yeah self user check is not implemented. But you could do that in the service directly

nickvergessen commented:
Earn bonus points for expanding https://github.com/nextcloud/server/blob/master/build/integration/features/provisioning-v1.feature

PVince81 committed:
Add known user check in avatar when v2-private scope

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Change account property capability

Include version number in capability

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

Co-authored-by: Joas Schilling <213943+nickvergessen@users.noreply.github.com>

PVince81 committed:
Int tests for provisioning API scopes

Added integration tests for the scope attributes in the provisioning
API.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
> Earn bonus points for expanding https://github.com/nextcloud/server/blob/master/build/integration/features/provisioning-v1.feature

done

nickvergessen reviewed (approved):
Wording and icons to discuss, but works technically

PVince81 committed:
Update autoloader for PlaceholderAvatar

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
not sure if related, might be the second time already:
```
/drone/src/tests/acceptance/features/app-files-sharing.feature:23
```

PVince81 commented:
not related, also happening on other PRs... could be a permanent thing on master...

PVince81 commented:
that test passes locally for me with :-(

PVince81 commented:
```
    And I accept the share for "/farewell.txt" in the notifications           # NotificationsContext::iAcceptTheShareForInTheNotifications()
      Notifications button in the header could not be found after 100 seconds (NoSuchElementException)
    And I open the Files app                                                  # FilesAppContext::iOpenTheFilesApp()
```

not sure what could delay the notification on CI

PVince81 commented:
still no green, that notification related test really doesn't want to pass any more: /drone/src/tests/acceptance/features/app-files-sharing.feature:23

MorrisJobke reviewed (approved):
Code looks good and scope changes work fine 👍

PVince81 commented:
rebased for CI

PVince81 committed:
Capability for federated scope

Added additional capability in the provisioning API to signal whether
the federation scope values can be used.

This is based on whether the lookup server upload is enabled or not.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
`/drone/src/tests/acceptance/features/app-files-sharing.feature:23` again ?! :frowning:

PVince81 commented:
can someone force merge maybe ? seems the notifications app might be cached or something... in any case, the test failure is unrelated

nickvergessen commented:
/backport to stable21
